cpp/src/io/json/json_tree.cu

```diff
@@ -439,210 +457,227 @@ rmm::device_uvector<size_type> hash_node_type_with_field_name(device_span<Symbol
 }
 
 /**
- * @brief Translates sorted parent_node_ids to parent_indices with indices from scatter_indices
- *
- * @param scatter_indices The sorted order of parent_node_ids
- * @param parent_node_ids The sorted parent_node_ids
- * @param stream CUDA stream used for device memory operations and kernel launches
- * @return Translated parent_indices pointing to sorted node_ids positions
- */
-rmm::device_uvector<NodeIndexT> translate_sorted_parent_node_indices(
-  device_span<size_type const> scatter_indices,
+ * @note Two level hashing algorithm
+ * 1. Convert node_category+fieldname to node_type. (passed as argument)
+ *   a. Create a hashmap to hash field name and assign unique node id as values.
+ *   b. Convert the node categories to node types.
+ *      Node type is defined as node category enum value if it is not a field node,
+ *      otherwise it is the unique node id assigned by the hashmap (value shifted by #NUM_CATEGORY).
+ * 2. Set operation on entire path of each node
+ *   a. Create a hash map with hash of {node_level, node_type} of its node and the entire parent
+ *      until root.
+ *   b. While creating hashmap, transform node id to unique node ids that are inserted into the
+ *      hash map. This mimicks set operation with hash map. This unique node ids are set ids.
+ *   c. Return this converted set ids, which are the hash map keys/values, and unique set ids.
+ **/
+std::pair<rmm::device_uvector<size_type>, rmm::device_uvector<size_type>> hash_node_path(
+  device_span<TreeDepthT const> node_levels,
+  device_span<size_type const> node_type,
   device_span<NodeIndexT const> parent_node_ids,
-  rmm::cuda_stream_view stream)
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  auto const num_nodes      = scatter_indices.size();
-  auto const gather_indices = cudf::detail::scatter_to_gather(
-    scatter_indices.begin(), scatter_indices.end(), num_nodes, stream);
+  auto num_nodes = parent_node_ids.size();
+  rmm::device_uvector<size_type> col_id(num_nodes, stream, mr);
```

Comment:
Many vars should be `auto const` if possible. This will be applied to all other places that I don't mention.

Reply:
To clarify: Any `auto` should accompany with `const` if possible.